code/cabal-audit/src/Distribution/Audit.hs

```diff
@@ -0,0 +1,246 @@
+import Security.Advisories.Cabal (ElaboratedPackageInfoAdvised, ElaboratedPackageInfoWith (MkElaboratedPackageInfoWith, elaboratedPackageVersion, packageAdvisories), matchAdvisoriesForPlan)
+import Security.Advisories.Convert.OSV qualified as OSV
+import Security.Advisories.Filesystem (listAdvisories)
+import System.Exit (exitFailure)
+import System.IO.Temp (withSystemTempDirectory)
+import System.Process (callProcess)
+import Validation (validation)
+
+data AuditException
+  = -- | parsing the advisory database failed
+    ListAdvisoryValidationError {originalFilePath :: FilePath, parseError :: [ParseAdvisoryError]}
+  | -- | to rethrow exceptions thrown by cabal during plan elaboration
+    CabalException {reason :: String, cabalException :: SomeException}
+  deriving stock (Show, Generic)
+
+instance Exception AuditException where
+  displayException = \case
+    ListAdvisoryValidationError dir errs ->
+      mconcat
+        [ "Listing the advisories in directory "
+        , dir
+        , " failed with: \n"
+        , show errs
+        ]
+    CabalException ctx (SomeException ex) ->
+      "cabal failed while "
+        <> ctx
+        <> ":\n"
+        <> displayException ex
+
+-- | the type of output that is chosen for the command
+data Output
+  = -- | write humand readable to stdout
+    HumanReadable
+  | -- | write as Osv format to the specified file
+    Osv {osvPath :: FilePath}
+
+-- | configuration that is specific to the cabal audit command
+data AuditConfig = MkAuditConfig
+  { advisoriesPathOrURL :: Either FilePath String
+  -- ^ path or URL to the advisories
+  , verbosity :: Verbosity.Verbosity
+  -- ^ verbosity of cabal
+  , outputType :: Output
+  }
+
+-- | the main action to invoke
+auditMain :: IO ()
+auditMain =
+  do
+    (auditConfig, nixStyleFlags) <- customExecParser (prefs showHelpOnEmpty) do
+      info (helper <*> auditCommandParser) do
+        mconcat
+          [ fullDesc
+          , progDesc (formatWith [blue] "audit your cabal projects for vulnerabilities")
+          , header (formatWith [bold, blue] "Welcome to cabal audit")
+          ]
+
+    buildAdvisories auditConfig nixStyleFlags
+      >>= handleBuiltAdvisories (outputType auditConfig)
+    `catch` \(SomeException ex) -> do
+      putStrLn $
+        unlines
+          [ formatWith [red, bold] "cabal-audit failed:"
+          , formatWith [red] $ displayException ex
+          ]
+      exitFailure
+
+buildAdvisories :: AuditConfig -> NixStyleFlags () -> IO (M.Map PackageName ElaboratedPackageInfoAdvised)
+buildAdvisories MkAuditConfig {advisoriesPathOrURL, verbosity} flags = do
+  let cliConfig = projectConfigFromFlags flags
+
+  ProjectBaseContext {distDirLayout, cabalDirLayout, projectConfig, localPackages} <-
+    establishProjectBaseContext
+      verbosity
+      cliConfig
+      OtherCommand
+      `catch` \ex -> throwIO $ CabalException {reason = "trying to establish project base context", cabalException = ex}
```

Comment:
Are you sure you want to catch async exceptions as well?

Reply:
I mean I don’t know where they’d be coming from…

Reply:
But I can rethrow them, I guess :)

Reply:
I'd suggest to use the package safe-exceptions instead.

Reply:
what would be the concrete benefit in this case?

Reply:
Not catching async exceptions, which is a very delicate matter and should be avoided, unless you absolutely know what you're doing.

Reply:
I understand that, where would an async exception come from though, we don't even compile this with `-threaded`

Reply:
I will rethrow async exceptions, I won't use `safe-exceptions`, though, afaict it doesn't do anything deeply magic that requires a dependency :)

Reply:
Yeah no I actually don’t know what the issue would be to handle these here, all of the exceptions we receive here would be exceptions thrown to the main thread, as this is the toplevel handler, we immediately terminate with exitFailure.

Reply:
It also doesn’t violate the rule “don’t recover from async exceptions” as exitFailure does throw an exception